### PR TITLE
replace_path property in the routing policies injected for backend APIs

### PR DESCRIPTION
**Pull request by guicassolato**
Closes [THREESCALE-3489](https://issues.jboss.org/browse/THREESCALE-3489)

### CODE REVIEW

File: app/lib/backend_api_logic/routing_policy.rb [hunk 1]
```diff
@@ -54,7 +54,7 @@ def as_json
                 value: path_to_regex
               ]
             }
-          }
+          }.merge(replace_path || {})
```
Comment (hallelujah) [inline app/lib/backend_api_logic/routing_policy.rb]:
```suggestion
          }.merge(replace_path)
```

File: app/lib/backend_api_logic/routing_policy.rb [hunk 2]
```diff
@@ -64,6 +64,13 @@ def path_to_regex
             "/#{path}/.*|/#{path}/?"
           end
         end
+
+        def replace_path
+          return if path.to_s.empty?
```
Comment (hallelujah) [inline app/lib/backend_api_logic/routing_policy.rb]:
```suggestion
          return {} if path.to_s.empty?
```

File: app/lib/backend_api_logic/routing_policy.rb [hunk 3]
```diff
@@ -64,6 +64,13 @@ def path_to_regex
             "/#{path}/.*|/#{path}/?"
           end
         end
+
+        def replace_path
+          return {} if path.to_s.empty?
```
Comment (hallelujah) [inline app/lib/backend_api_logic/routing_policy.rb]: Is it a case where we need to cast it with `to_s`?

```suggestion
          return {} if path.blank?
```

File: test/unit/backend_api_logic/routing_policy_test.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
+# frozen_string_literal: true
+
+require 'test_helper'
+
+module BackendApiLogic
+  class RoutingPolicyTest < ActiveSupport::TestCase
+    setup do
+      @service = FactoryBot.create(:service)
```
Comment (Martouta) [inline test/unit/backend_api_logic/routing_policy_test.rb]:
```suggestion
      @service = FactoryBot.create(:simple_service)
```

The only difference between those 2 factories is that the account it creates is simple or with a bunch of stuff, and in this test you don't seem to want to use the account at all so it seems fair enough to make it as efficient as possible then :smile:

File: test/unit/backend_api_logic/routing_policy_test.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+    test '#with_subpaths?' do
+      assert proxy.with_subpaths?
+      other_proxy = FactoryBot.create(:proxy)
+      refute other_proxy.with_subpaths?
+    end
+
+    test '#policy_chain' do
+      injected_rules = [
+        { url: service.backend_apis.last.private_endpoint,  condition: { operations: [match: :path, op: :matches, value: '/foo/.*|/foo/?'] }, replace_path: '{{original_url.path | replace: "/foo", "/"}}' },
```
Comment (Martouta) [inline test/unit/backend_api_logic/routing_policy_test.rb]: Check this PR: https://github.com/3scale/porta/pull/1206
In this case, for `service.backend_apis.last` and the next line `service.backend_apis.first`, the order will be random for Postgres and sometimes we will get random CircleCI failures for that :disappointed: So you can fix it by something like:
```ruby
injected_rules = [
  { url: service.backend_apis.order(:id).last.private_endpoint,  condition: { operations: [match: :path, op: :matches, value: '/foo/.*|/foo/?'] }, replace_path: '{{original_url.path | replace: "/foo", "/"}}' },
  { url: service.backend_apis.order(:id).first.private_endpoint, condition: { operations: [match: :path, op: :matches, value: '/.*'] } }
]
```

Comment (guicassolato) [reply test/unit/backend_api_logic/routing_policy_test.rb]: Taken care by @hallelujah already: https://github.com/3scale/porta/blob/83a66a93c9bf04a3ae383327e111c324bd1b8096/app/models/backend_api_config.rb#L6

File: app/lib/backend_api_logic/routing_policy.rb [hunk 6]
```diff
@@ -64,6 +64,13 @@ def path_to_regex
             "/#{path}/.*|/#{path}/?"
           end
         end
+
+        def replace_path
+          return {} if path.blank?
+
+          private_base_path = URI.parse(private_endpoint).path.to_s.sub(/^\//, '').sub(/\/$/, '')
```
Comment (hallelujah) [inline app/lib/backend_api_logic/routing_policy.rb]: You might want to use https://github.com/3scale/porta/blob/83a66a93c9bf04a3ae383327e111c324bd1b8096/app/lib/string_utils/strip_slash.rb#L8

Comment (guicassolato) [reply app/lib/backend_api_logic/routing_policy.rb]: Hmmm. It should not be `gsub`.

**Pull request merged**